File: leetcode3/MinimumCostHomecomingofaRobotinaGrid/a.cpp
/* Copyright (C) 2022 by iamslash */

#include <cstdio>
#include <vector>

using namespace std;

// 252ms 55.82% 149.9MB 26.89%
// matrix
// O(N) O(1)
class Solution {
public:
  int minCost(vector<int>& startPos, vector<int>& homePos, vector<int>& rowCosts, vector<int>& colCosts) {
    int cost = 0, y = startPos[0], x = startPos[1], z = homePos[0], w = homePos[1];
    while (y != z) {
      y += (z - y) / abs(z - y);
      cost += rowCosts[y];
    }
    while (x != w) {
      x += (w - x) / abs(w - x);
      cost += colCosts[x];
    }
    return cost;
  }
};

int main() {
  return 0;
}

File: leetcode3/MinimumCostHomecomingofaRobotinaGrid/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2022 by iamslash

# 1676ms 76.98% 28.9MB 50.86%
# matrix
# O(N) O(1)
class Solution:
    def minCost(self, startPos: List[int], homePos: List[int], rowCosts: List[int], colCosts: List[int]) -> int:
        cost, y, x, z, w = 0, startPos[0], startPos[1], homePos[0], homePos[1]
        while y != z:
            y += int((z - y) / abs(z - y))
            cost += rowCosts[y]
        while x != w:
            x += int((w - x) / abs(w - x))
            cost += colCosts[x]
        return cost
